lib/devicetrust/native/device_windows.go

```diff
@@ -0,0 +1,489 @@
+	}
+
+	return ak, nil
+}
+
+func enrollDeviceInit() (*devicepb.EnrollDeviceInit, error) {
+	akPath, err := setupDeviceStateDir(os.UserHomeDir)
+	if err != nil {
+		return nil, trace.Wrap(err, "setting up device state directory")
+	}
+
+	tpm, err := openTPM()
+	if err != nil {
+		return nil, trace.Wrap(err, "opening tpm")
+	}
+	defer tpm.Close()
+
+	// Try to load an existing AK in the case of re-enrollment, but, if the
+	// AK does not exist, create one and persist it.
+	ak, err := loadAK(tpm, akPath)
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err, "loading ak")
+		}
+		log.Debug("No existing AK was found on disk, an AK will be created.")
+		ak, err = createAndSaveAK(tpm, akPath)
+		if err != nil {
+			return nil, trace.Wrap(err, "creating ak")
+		}
+	} else {
+		log.Debug("Existing AK was found on disk, it will be reused.")
+	}
+	defer ak.Close(tpm)
+
+	deviceData, err := collectDeviceData()
+	if err != nil {
+		return nil, trace.Wrap(err, "collecting device data")
+	}
+
+	marshaledEK, err := getMarshaledEK(tpm)
+	if err != nil {
+		return nil, trace.Wrap(err, "marshalling ek")
+	}
+
+	credentialID, err := credentialIDFromAK(ak)
+	if err != nil {
+		return nil, trace.Wrap(err, "determining credential id")
+	}
+
+	return &devicepb.EnrollDeviceInit{
+		CredentialId: credentialID,
+		DeviceData:   deviceData,
+		Tpm: &devicepb.TPMEnrollPayload{
+			Ek: &devicepb.TPMEnrollPayload_EkKey{
+				EkKey: marshaledEK,
+			},
+			AttestationParameters: devicetrust.AttestationParametersToProto(
+				ak.AttestationParameters(),
+			),
+		},
+	}, nil
+}
+
+// credentialIDFromAK produces a deterministic short-ish unique-ish human
+// readable-ish string identifier for a given AK. This can then be used as a
+// reference for this AK in the backend.
+//
+// To produce this, we perform a SHA256 hash over the constituent fields of
+// the AKs public key and then base64 encode it to produce a human-readable
+// string. This is similar to how SSH fingerprinting of public keys work.
+func credentialIDFromAK(ak *attest.AK) (string, error) {
+	akPub, err := attest.ParseAKPublic(
+		attest.TPMVersion20,
+		ak.AttestationParameters().Public,
+	)
+	if err != nil {
+		return "", trace.Wrap(err, "parsing ak public")
+	}
+	publicKey := akPub.Public
+	switch publicKey := publicKey.(type) {
+	case *rsa.PublicKey:
+		h := sha256.New()
+		// This logic is roughly based off the openssh key fingerprinting,
+		// but, the hash excludes "ssh-rsa" and the outputted id is not
+		// prepended with "SHA256:
+		//
+		// It is imperative the order of the fields does not change in future
+		// implementations.
+		h.Write(big.NewInt(int64(publicKey.E)).Bytes())
+		h.Write(publicKey.N.Bytes())
+		return base64.RawStdEncoding.EncodeToString(h.Sum(nil)), nil
+	default:
+		return "", trace.BadParameter("unsupported public key type: %T", publicKey)
+	}
+}
+
+// getDeviceSerial returns the serial number of the device using PowerShell to
```

Comment:
A bit of a silly Q, but can we count on PowerShell always being available? Would this require a min Windows version/patch/build?

How harder would be to make this without exec-ing anything else?

Reply:
PowerShell has been shipped pre-installed since Windows 7 SP1.

There's roughly two paths for us fetching this information without PowerShell:
- Query WMI objects through the COM API
- Find Win32 API calls that provide this information without WMI

The first is probably the more "sanctioned" way of doing so without PowerShell, but is still relatively unpleasant from Go. A package exists for this (https://github.com/yusufpapurcu/wmi) but it's not an extremely popular or well-documented package and is itself mostly dependent on another package (https://github.com/go-ole/go-ole) which has not been updated in some 8 years. My major concern is that this package is not extremely popular, and in the case that we experience an issue, it's going to require knowledge of the Windows API that we do not possess internally in order to fix it. I still feel extremely out of depth in this area, and certainly wouldn't be comfortable trying to account for edge-cases with the Windows API that I'm not aware of.

For some information, it's also possible to find Win32 API calls that provide it without going through WMI. This is pretty obscure and seems to require a lot of knowledge of what libraries Win32 offers, and I'm only able to find a few of the fields we need through googling and the rest appear to be not available. This is also potentially more awkward than querying WMI, and I'm unsure what the backwards/forwards compatibility on these APIs is.

It's also looking like we are going to be requiring PowerShell in order to provide the UAC pop-up on Enrollment.

Therefore, my gut feeling, is that we should ship with a dependency on PowerShell, and if we find a compelling enough argument down the road to switch to invoking WMI directly through COM APIs then we can do so then.

Reply:
Thanks for the detailed explanation, sounds good.